packages/redux-routine/src/index.js

```diff
@@ -0,0 +1,51 @@
+/**
+ * Internal dependencies
+ */
+import isGenerator from './is-generator';
+
+/**
+ * Creates a Redux middleware, given an object of controls where each key is an
+ * action type for which to act upon, the value a function which returns either
+ * a promise which is to resolve when evaluation of the action should continue,
+ * or a value. The value or resolved promise value is assigned on the return
+ * value of the yield assignment. If the control handler returns undefined, the
+ * execution is not continued.
+ *
+ * @param {Object} controls Object of control handlers.
+ *
+ * @return {Function} Redux middleware function.
+ */
+export default function createMiddleware( controls = {} ) {
+	return ( store ) => ( next ) => ( action ) => {
+		if ( ! isGenerator( action ) ) {
+			return next( action );
+		}
+
+		function step( nextAction ) {
+			if ( ! nextAction ) {
+				return;
+			}
+
+			const control = controls[ nextAction.type ];
+			if ( typeof control === 'function' ) {
+				const routine = control( nextAction );
```

Comment:
I think the `control` should also receive a function to raise errors and sometimes you want to avoid the recursion and just yields the value directly.

Question: Does it support yielding generators inside generators. I think this should be built-in to encourage composition?

Reply:
>I think the `control` should also receive a function to raise errors and sometimes you want to avoid the recursion and just yields the value directly.

Could potentially have it so that the control could reject its promise. I'm not sure what impact this should have though; I wanted to stay as unopinionated as possible, leaving this pattern to user-space.

Reply:
>Question: Does it support yielding generators inside generators. I think this should be built-in to encourage composition?

I expect it should work out of the box, yes, since it's just a middleware. Should have a unit test for this though.

Reply:
Okay, so I see how this is handled in `rungen` is to throw as an error when a "reject" state is reached, allowing the developer to handle as `try` / `catch` in their action creator. Seems like a reasonable approach.